src/par_iter/for_each_atomic/flat_combine.rs

```diff
@@ -0,0 +1,230 @@
+use thread_pool::WorkerThread;
+use util::PoisonPanic;
+
+pub struct FlatCombiner<F, T>
+    where T: Send,
+          F: FnMut(T)
+{
+    senders: Vec<Sender<T>>,
+    owner_data: Mutex<OwnerData<F>>,
+}
+
+struct Sender<T>
+    where T: Send
+{
+    data: AtomicPtr<T>
+}
+
+struct OwnerData<F> {
+    func: F,
+}
+
+#[must_use]
+struct BecameOwner(bool);
+
+impl BecameOwner {
+    fn became_owner(&self) -> bool { self.0 }
+}
+
+impl<F, T> FlatCombiner<F, T>
+    where T: Send,
+          F: FnMut(T)
+{
+    /// Create a `FlatCombiner`.
+    ///
+    /// Unsafe because this must be called from inside a worker thread
+    /// in the same registry as the one from which the later calls to
+    /// `produce()` will occur.
+    #[allow(dead_code)] // FIXME
+    pub unsafe fn new(func: F) -> FlatCombiner<F, T> {
+        let worker = WorkerThread::current();
+        debug_assert!(!worker.is_null());
+        let num_threads = (*worker).num_threads();
+
+        FlatCombiner {
+            senders: (0..num_threads).map(|_| Sender::new()).collect(),
+            owner_data: Mutex::new(OwnerData { func: func }),
+        }
+    }
+
+    /// Executes the closure on `data` and returns. This executes
+    /// atomically with respect to all other calls to `produce()`.
+    ///
+    /// Unsafe because this must be called from inside a worker thread
+    /// in the same registry as the one from which the
+    /// `FlatCombiner::new()` method was invoked.
+    pub unsafe fn produce(&self, data: T) {
+        let worker = WorkerThread::current();
+        debug_assert!(!worker.is_null());
+
+        let data = Some(data); // meh
+
+        let index = (*worker).index();
+        let sender = &self.senders[index];
+
+        // We have not yet enqueued any data, so there can't be any.
+        debug_assert!(sender.is_empty());
+
+        // Otherwise, enqueue the data.
+        self.send(sender, data.unwrap());
+    }
+
+    unsafe fn send(&self,
+                   sender: &Sender<T>,
+                   mut data: T)
+    {
+        // At this point, the data is owner by the queue, so forget
+        // it. The `data_guard` that is returned by the queue will
+        // ensure we free the data if a panic should occur.
+        let data_ptr: *mut T = &mut data;
+        mem::forget(data);
+        let data_guard = sender.put(data_ptr);
+
+        loop {
+            if self.try_to_become_owner().became_owner() {
+                // If we became owner, then all data
+                // would be processed, and so `sender.is_empty()`
+                // must be true, and the `data_guard` is not needed.
+                debug_assert!(sender.is_empty());
+                mem::forget(data_guard);
+                return;
+            }
+
+            if sender.is_empty() {
+                // Our request was served, so we can just return, and
+                // `data_guard` is not needed.
+                mem::forget(data_guard);
+                return;
```

Comment:
I think you could get around this with separate reads and writes, instead of `try_take`'s atomic swap.  Read the pointer, and if it's not null work on it.  No one else will race here because there's only one global owner.  Null the pointer only after you're done to indicate "request was served".